Source/WebCore/platform/graphics/PixelBuffer.h

```diff
@@ -39,91 +38,33 @@ namespace WebCore {
 class PixelBuffer : public RefCounted<PixelBuffer> {
     WTF_MAKE_NONCOPYABLE(PixelBuffer);
 public:
-    static bool supportedPixelFormat(PixelFormat);
+    WEBCORE_EXPORT static bool supportedPixelFormat(PixelFormat);
 
-    WEBCORE_EXPORT static RefPtr<PixelBuffer> tryCreate(const PixelBufferFormat&, const IntSize&);
-    WEBCORE_EXPORT static RefPtr<PixelBuffer> tryCreate(const PixelBufferFormat&, const IntSize&, Ref<JSC::ArrayBuffer>&&);
-    WEBCORE_EXPORT static Ref<PixelBuffer> create(const PixelBufferFormat&, const IntSize&, JSC::Uint8ClampedArray&);
-
-    WEBCORE_EXPORT ~PixelBuffer();
+    WEBCORE_EXPORT virtual ~PixelBuffer();
 
     const PixelBufferFormat& format() const { return m_format; }
     const IntSize& size() const { return m_size; }
-    JSC::Uint8ClampedArray& data() const { return m_data.get(); }
 
-    Ref<JSC::Uint8ClampedArray>&& takeData() { return WTFMove(m_data); }
+    virtual uint8_t* bytes() const = 0;
```

Comment:
Every call to item() now calls the bytes() virtual function. I think this would be a measurable performance hit for filter effect code that can do this inside tight loops.

Also since item() is exported, it won't be inlined.

Can we fetch the start pointer once via a virtual function at creation time, and after that just index off it?

Reply:
You are right. 

I changed ByteArrayPixelBuffer to pass the raw data pointer and the sizeInBytes to the PixelBuffer constructor and I made the methods `bytes() `and `sizeInBytes()` non virtual. I think I can't call virtual methods from the base class constructor.